WinImageBuilder.psm1

```diff
@@ -409,20 +531,213 @@ function Add-VirtIODriversFromISO($vhdDriveLetter, $image, $isoPath) {
-{
+function Compress-Image {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$VirtualDiskPath,
+        [Parameter(Mandatory=$true)]
+        [string]$ImagePath
+    )
+    if (!(Test-Path $VirtualDiskPath)) {
+        Throw "$VirtualDiskPath not found"
+    }
+    $tmpName = $ImagePath + "." + (Get-Random)
+
+    $7zip = Join-Path $localResourcesDir 7za.exe
+    $pigz = Join-Path $localResourcesDir pigz.exe
+    try {
+        Write-Host "Archiving $VirtualDiskPath to tarfile $tmpName"
+        pushd ([System.IO.Path]::GetDirectoryName((Resolve-path $VirtualDiskPath).Path))
+        try {
+            # Avoid storing the full path in the archive
+            $imageFileName = (Get-Item $VirtualDiskPath).Name
+            Write-Host "Creating tar archive..."
+            & $7zip a -ttar $tmpName $imageFileName
+            if ($LASTEXITCODE) {
+                if ((Test-Path $imageFileName)) {
+                    Remove-Item -Force $imageFileName
+                }
+                throw "7za.exe failed while creating tar file for image: $tmpName"
+            }
+        } finally {
+            popd
+        }
+
+        Remove-Item -Force $VirtualDiskPath
+        Write-Host "Compressing $tmpName to gzip"
+        pushd ([System.IO.Path]::GetDirectoryName((Resolve-path $tmpName).Path))
+        try {
+            $tmpPathName = (Get-Item $tmpName).Name
+            Write-Host "Creating gzip..."
+            & $pigz -p12 $tmpPathName
+            if ($LASTEXITCODE) {
+                $gzipped = ($tmpPathName + ".gz")
+                if ((Test-Path $gzipped)) {
+                    Remove-Item -Force $gzipped
+                }
+                throw "pigz.exe failed while creating gzip file for : $tmpName"
+            }
+        } finally {
+            popd
+        }
+    } catch {
+        Remove-Item -Force $tmpName -ErrorAction SilentlyContinue
+        Remove-Item -Force $VirtualDiskPath -ErrorAction SilentlyContinue
+        throw
+    }
+    Move-Item ($tmpName + ".gz") $ImagePath
+    Write-Output "MaaS image is ready and available at: $ImagePath"
+}
+
+function Shrink-VHDImage {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$VirtualDiskPath
+    )
+    Write-Host "Shrinking VHD to minimum size"
+
+    $vhdSize = (Get-VHD -Path $VirtualDiskPath).Size
+    $vhdSizeGB = $vhdSize/1GB
+    Write-Host "Initial VHD size is: $vhdSizeGB GB"
+
+    $Drive = (Mount-VHD -Path $VirtualDiskPath -Passthru | Get-Disk | Get-Partition | Get-Volume).DriveLetter
+    Optimize-Volume -DriveLetter $Drive -Defrag -ReTrim
+
+    $partitionInfo = Get-Partition -DriveLetter $Drive
+    $MinSize = (Get-PartitionSupportedSize -DriveLetter $Drive).SizeMin
+    $CurrSize = ((Get-Partition -DriveLetter $Drive).Size/1GB)
+    Write-Host "Current partition size: $CurrSize GB"
+    # Leave at least 500MB for making sure Sysprep finishes successfuly
+    $newSizeGB = [int](($MinSize + 500MB)/1GB) + 1

```

Comment:
What happens if $MinSize is close to the maximum size of the VHD? Say, I created a VHD that is smaller then it should be. After the OS gets installed and sysprepped, I am left with 200 MB free space. If I set the $newSizeGB to $MinSize + 500MB, won't that make the new partition size larger then the actual VHD? If so, should we resize the VHD as well to a larger size?